### DISCUSSION TITLE
How to Read BuildKit Build Logs

### DISCUSSION THREAD

**Discussion by dfoster42** [moby/moby#43788] [Q&A] [answered]
Howdy.  I'm working on a cli tool that builds docker images as part of its workflow

It was relatively straightforward serving the logs with the v1 builder, but I'm having a hard time figuring out how to serve the logs to the end user when using builderkit.

```
	buildOpts := types.ImageBuildOptions{
		Version: types.BuilderBuildKit,
		Tags: []string{
			containerImage + ":" + containerVersion,
			containerImage + ":latest",
		},
		BuildArgs: buildArgs,
	}

	buildOutput, err := r.GetClient().ImageBuild(context.Background(), buildCtx, buildOpts)
```

Before specifying version, we were just reading json logs from `buildOutput.Body`, but now that only serves the `Successfully tagged` message at the end.

Could someone help understand how to visualize the builderkit build logs like you would the v1 builder logs?

**Comment by thaJeztah** [comment]
BuildKit client is a bit more involved than the classic builder; the code in the Docker CLI could be a starting point; https://github.com/docker/cli/blob/v20.10.17/cli/command/image/build_buildkit.go

However, for the next release, build with buildkit will be backed by builds, which implements more of the BuildKit features; code can be found here; https://github.com/docker/buildx/blob/5b452b72a2bf9945bb7d062cf28842430f910777/commands/build.go#L90

**Reply by dfoster42** [reply]
Thank you for the info!  Apologies in advance for my ignorance.  Regarding the next release: does this mean we'll be able to simply call client.ImageBuild, and read from the output's body for build logs as we can without buildkit, or will we need to hook into other api calls  or something like that?

**Comment by matt-allan** [comment, marked as answer]
This is what I have figured out so far:

You need to use the jsonmessage` package (`github.com/docker/docker/pkg/jsonmessage`) to read the response body like you did with V1.

I think there is a helper function in that package you can use. I use a custom function that looks like this:

```go
func DockerMessages(in io.Reader, auxCallback func(jsonmessage.JSONMessage)) error {
	dec := json.NewDecoder(in)

	for {
		var jm jsonmessage.JSONMessage
		if err := dec.Decode(&jm); err != nil {
			if err == io.EOF {
				break
			}
			return fmt.Errorf("decode JSON message: %w", err)
		}

		// If the message contains an error, return it to the caller
		// after writing. This is how Docker communicates things like
		// image builds failing.
		if jm.Error != nil {
			return jm.Error
		}

		// Check for out of band data (like image IDs, digests, etc.)
		if jm.Aux != nil {
			if auxCallback != nil {
				auxCallback(jm)
			}
			continue // (docker ignores the rest if there is aux data)
		}
		
		// ...
```

With Buildkit all of the data is on `JSONMessage.Aux`.  You can switch on `JSONMessage.ID` to guess what the Aux type will be. For buildkit output from an image build it will be `moby.buildkit.trace`.

The buildkit trace is a binary encoded protobuf message which is then JSON encoded, so it's a base64 JSON string. You can decode it like this:

```go
import controlapi "github.com/moby/buildkit/api/services/control"

var bytes []byte
if err := json.Unmarshal(*jm.Aux, &bytes); err != nil {
	panic(err)
}
resp := controlapi.StatusResponse{}
if err := proto.Unmarshal(bytes, &resp); err != nil {
	panic(err)
}
```

The proto is defined [here](https://github.com/moby/buildkit/blob/326d505d71c2b86d0735cd0a38b22b34c84aec68/api/services/control/control.proto#L117). 

Edit: here's a more complete example that just prints the output to the terminal: https://gist.github.com/matt-allan/9d5419abd2c234011de175e891be48fb

It uses the "github.com/moby/buildkit/util/progress/progressui" package to print the output.

**Comment by Das-Rabindra** [comment]
Parse JSON messages from buildOutput.Body as a stream:
Instead of reading plain text logs, you need to decode each JSON message. These are usually structs like:

`type BuildkitStatus struct {
  Stream string `json:"stream,omitempty"`
  Aux    struct {
    ID string `json:"ID,omitempty"`
  } `json:"aux,omitempty"`
  Error string `json:"error,omitempty"`
}
`
example `decoder := json.NewDecoder(buildOutput.Body)
for {
    var msg map[string]interface{}
    if err := decoder.Decode(&msg); err != nil {
        if err == io.EOF {
            break
        }
        log.Fatalf("Error decoding build output: %v", err)
    }
    if stream, ok := msg["stream"]; ok {
        fmt.Print(stream)
    }
    if errMsg, ok := msg["error"]; ok {
        fmt.Fprintf(os.Stderr, "Build error: %v\n", errMsg)
    }
}
`
